src/parsers/manifest/metaplaylist/index.ts

```diff
@@ -0,0 +1,309 @@
+  contents: Array<{
+    url: string;
+    startTime: number;
+    endTime: number;
+    transport: string;
+    textTracks: [{
+      url: string;
+      language: string;
+      mimeType: string;
+    }];
+  }>;
+  attributes: {
+    timeShiftBufferDepth? : number;
+    dynamic? : boolean;
+  };
+}
+
+/**
+ * Parse playlist string to JSON.
+ * Returns an array of contents.
+ * @param {string} data
+ */
+export default function parseMetaPlaylist(
+  data: string,
+  url : string
+): IParserResponse<IParsedManifest> {
+  let parsedData;
+  try {
+    parsedData = JSON.parse(data);
+  } catch (error) {
+    throw new Error("MPL Parser: Bad MetaPlaylist file. Expected JSON.");
+  }
+
+  const { contents } = parsedData;
+
+  // quick checks
+  if (contents == null || contents.length === 0) {
+    throw new Error("MPL Parser: No content found.");
+  }
+  if (parsedData.attributes == null) {
+    throw new Error("MPL Parser: Malformed playlist.");
+  }
+  const ressources : Array<{ url : string; transportType : string }> = [];
+  for (let i = 0; i < contents.length; i++) {
+    const content = contents[i];
+    if (
+      content.url == null ||
+      content.startTime == null ||
+      content.endTime == null ||
+      content.transport == null
+    ) {
+      throw new Error("MPL Parser: Malformed content.");
+    }
+    ressources.push({ url: content.url, transportType: content.transport });
+  }
+
+  const metaPlaylist : IMetaPlaylist = parsedData;
+  return {
+    type : "needs-manifest-loader",
+    value : {
+      ressources,
+      continue : function parseWholeMPL(loadedRessources : IParsedManifest[]) {
+        const parsedManifest = createManifest(metaPlaylist, loadedRessources, url);
+        return { type: "done", value: parsedManifest };
+      },
+    },
+  };
+}
+
+/**
+ * From several parsed manifests, generate a single manifest
+ * which fakes live content playback.
+ * Each content presents a start and end time, so that periods
+ * boudaries could be adapted.
+ * @param {Object} contents
+ * @param {string} url
+ */
+function createManifest(
+  mplData: IMetaPlaylist,
+  manifests : IParsedManifest[],
+  url: string
+): IParsedManifest {
+  const { contents } = mplData;
+  const endOfContent = contents[contents.length - 1].endTime;
+
+  const periods : IParsedPeriod[] = [];
+  for (let iMan = 0; iMan < contents.length; iMan++) {
+    const content = contents[iMan];
+    const periodOffset = content.startTime;
+    const currentManifest = manifests[iMan];
+    for (let iPer = 0; iPer < currentManifest.periods.length; iPer++) {
+      const currentPeriod = currentManifest.periods[iPer];
+      const adaptationTypes = Object.keys(currentPeriod.adaptations);
+
+      const adaptations = adaptationTypes
+        .reduce((acc : IParsedAdaptations, type : string) => {
+          const currentAdaptations = currentPeriod.adaptations[type];
+          if (currentAdaptations == null) {
+            return acc;
+          }
```

Comment:
![Bug](https://sonarsource.github.io/sonar-github/bug.png 'Bug') Bug: Correct one of the identical sub-expressions on both sides of operator "&&" ([typescript:S1764](https://sonarcloud.io/organizations/rx-player/rules#rule_key=typescript%3AS1764))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=rx-player&pullRequest=367&issues=AWd1fPma8YaV-M4wPwBi&open=AWd1fPma8YaV-M4wPwBi)

[ISSUE_KEY]: AWd1fPma8YaV-M4wPwBi